### DISCUSSION TITLE
using our own auth?

### DISCUSSION THREAD

**Discussion by fergusmeiklejohn** [supabase/supabase#3899] [Questions]
How painful would it be to use our own authentication, while integrating with the rest of Supabase?
Reason: Supabase auth is client based, so to generate a cookie server side we need to use the onAuthChanged listener client side. If we're making a server side app, it would be better to handle this on the server. I can roll my own auth but worry that this would make integration with the rest of supabase more painful.. What steps would be involved to use our own user authentication? Thanks :-)

**Comment by awalias** [comment]
thanks for the question @fergusmeiklejohn, I don't think I fully understand what the use case is?

Do your users still log in somewhere with email + password?

some considerations that may be useful:

have you looked into supabase-js's [setAuth()](https://supabase.io/docs/reference/javascript/auth-setauth) method?

in some circumstances you can use the gotrue jwt_secret to mint your own jwts using any jwt lib, you can include custom claims, and then write RLS policies to check your new claims etc.

**Reply by fergusmeiklejohn** [reply]
I think what we need is for supabase to support OAuth2: https://github.com/supabase/supabase/discussions/417#discussioncomment-770022
That would enable us to do auth as we need to while fully integrating into the rest of Supabase.

**Comment by mstade** [comment]
For what it's worth, for our app [Paperworker](https://www.paperworker.se) we have deployed our own oauth provider. The main reason for this is that we don't use passwords for authentication, we use something called [BankID](https://www.bankid.com/en/) which is an identity system deployed by a conglomerate of Swedish banks. The way we do this is by minting our own JWT tokens using the secret available in the dashboard, as suggested by @awalias above. Another reason we decided to deploy our own oauth provider is that we need a fully native authentication flow, since our client is a mobile app and we don't want/need the whole browser flow to deal with login/consent screens.

This works well, but frankly we're not entirely happy about the situation, mainly because there's no way as far as we know to rotate the JWT secret. This means if our identity provider is compromised we don't have a good way to ensure the integrity of the system other than to destroy and recreate the environment anew. We've considered implementing a function in Supabase that would mint the tokens instead, such that we could make our identity provider just be a Supabase user with a special role that has access to mint JWT tokens on behalf of other users, but we have yet to look into that solution any further.

Any official guidance on this would be much appreciated!

**Reply by awalias** [reply]
this minting function is a nice idea, it would improve DX for everyone wanting to sign custom JWTs, I've started an issue to track it here:

https://github.com/supabase/gotrue/issues/277

thoughts and code contributions always welcome! 🙏

**Reply by awalias** [reply]
also noted on your point re: secret rolling, will make this high priority item

**Reply by mstade** [reply]
To be clear, I think the minting function is a _workaround_ for our case, because of the lack of either:

1. A generic OpenID Connect adapter that would let us point to _any_ compliant OIDC provider
2. An identity provider solution deployed as part of the supabase package, that would also allow for custom interactions for things like login and consent screens

The latter is a pretty big undertaking and also requires pretty specific skillsets not just to deploy but also maintain – it's probably expensive enough to warrant making a startup of its own. The former is an adapter just like any of the other social ones, just that it's generic to any compliant OIDC provider (e.g. AWS Cognito, Auth0, Ory Hydra) and would really just require pointing to the provider's domain to find its `/.well-known/openid-configuration` document that tells how to configure the provider.

Regardless of whether there's a JWT minting function or not, having the ability to regularly and even automatically roll over the JWT secret would be very appreciated. It does introduce some fun cases though, like if someone logs in right before the JWT secret is rolled over to a new one. The token verification mechanism would have to deal with verifying both against the new secret and one (or more) older secrets, potentially with the need to blacklist specific secrets in case there's a risk that it's been compromised. Another thing to consider is whether the token signing should really be done using an asymmetric method rather than HMAC.

Thank you for raising the issue, I'll try to contribute best I can!

**Comment by sergiodxa** [comment]
To use Supabase without including the client-side lib as @fergusmeiklejohn mentioned, the service should expose OAuth2 endpoints, so it needs an authorization, token and profile endpoint.

That would allow apps using server-side frameworks to use Supabase for auth, when redirecting to the authorization endpoint the user should see the authentication options configured in the Supabase account (e.g. what social networks to use).

After the OAuth2 flow is completed the app should have a token to use to talk with Supebase in behalf of the user.

**Comment by dinvlad** [comment]
Could you clarify if Supabase.io (managed) service exposes a `.well-known/openid-configuration` endpoint for the token issuer?

EDIT: It looks like this is currently not possible due to the use of symmetric signing algo (`HS256`), but there're [plans ](https://github.com/supabase/supabase/discussions/5911#discussioncomment-2371302)to support it.

**Reply by mb21** [reply]
current status of adding RS256 JWT Signing (i.e. asymmetric/public key cryptography): https://github.com/orgs/supabase/discussions/12759
